src/components/native-block-inserter-button/index.jsx

```diff
@@ -80,46 +82,122 @@ export default function NativeBlockInserterButton() {
+			return true;
+		} catch ( error ) {
+			debug( 'Failed to insert block:', error );
+			return false;
+		}
+	};
+
+	/**
+	 * Insert media blocks from native media picker using block transforms.
+	 *
+	 * This method uses the same file transform system as drag-and-drop,
+	 * ensuring consistent behavior and leveraging WordPress's extensibility:
+	 * - Finds matching transform based on file types (image/video/audio)
+	 * - Respects transform priorities (gallery > single image)
+	 * - Supports third-party block transforms
+	 * - Handles block insertion validation
+	 *
+	 * @param {Array} mediaArray Array of media objects with shape:
+	 *                           { id?, url, type, caption?, alt?, title?, metadata? }
+	 * @return {Promise<boolean>} True if insertion succeeded, false otherwise
+	 */
+	const insertMedia = async ( mediaArray ) => {
+		if ( ! Array.isArray( mediaArray ) || mediaArray.length === 0 ) {
+			return false;
+		}
+		try {
+			// Convert media objects to File objects
+			// This allows us to use the existing file transform system
+			const files = await Promise.all(
+				mediaArray.map( async ( media ) => {
+					try {
+						const response = await fetch( media.url );
+						const blob = await response.blob();
+						const filename = media.url.split( '/' ).pop() || 'media';
+						return new File( [ blob ], filename, {
+							type: media.type ?? 'application/octet-stream',
+						} );
+					} catch ( error ) {
+						debug(
+							`insertMedia: Failed to fetch media: ${ media.url }`,
+							error
+						);
+						return null;
+					}
+				} )
+			);
+
+			// Filter out any failed fetches
+			const validFiles = files.filter( ( f ) => f !== null );
+
+			if ( validFiles.length === 0 ) {
+				debug( 'insertMedia: No valid files to insert' );
+				return false;
+			}
+
+			// Find matching transform using WordPress's transform system
+			// This is the same logic as onFilesDrop in use-on-block-drop
+			const transformation = findTransform(
+				getBlockTransforms( 'from' ),
+				( transform ) =>
+					transform.type === 'files' &&
+					transform.isMatch &&
+					transform.isMatch( validFiles )
+			);
+
+			if ( ! transformation ) {
+				debug( 'insertMedia: No matching transform found', {
+					fileCount: validFiles.length,
+					fileTypes: validFiles.map( ( f ) => f.type ),
+				} );
+				return false;
+			}
+
+			const blocks = transformation.transform(
+				validFiles,
+				updateBlockAttributes
+			);
+
+			if (
+				! blocks ||
+				( Array.isArray( blocks ) && blocks.length === 0 )
+			) {
+				debug( 'insertMedia: Transform produced no blocks' );
+				return false;
+			}
+
+			onInsertBlocks( blocks );
+			return true;
+		} catch ( error ) {
+			debug( 'insertMedia: Failed to insert media blocks', error );
+			return false;
+		}
+	};
 
 	return (
 		<Button
 			title={ __( 'Add block' ) }
 			icon={ plus }
 			onClick={ ( e ) => {
 				e.preventDefault();
+
+				const blocks = serializeBlocksForNative( inserterItems );
```

Comment:
This is the fix for [CMM-903: Performance: Inserter recomputes the list of blocks on every change in the editor](https://linear.app/a8c/issue/CMM-903/performance-inserter-recomputes-the-list-of-blocks-on-every-change-in). Instead of `useEffect`, it now serializes blocks only when it's needed.

Ideally, I would also ensure `useInsertionPoint` isn't called so frequently, but it could be done later.

Reply:
Sounds good. We can revisit other [memoization ideas](https://github.com/wordpress-mobile/GutenbergKit/pull/199#discussion_r2474322561) later. Maybe we can further reduce `useInsertionPoint` invocations.